Add To API Query With Multiple Criteria

SpiritBreaker226 opened the pull request:
Before you merge this into master let me know if there is any combines of data you would like to use on this pull request.

Data you can get in any combines you would like using those attributes
## Councillors

_ward_id_ - Display the councillor who is representing this ward
_counciillor_name_ - Uses the the councillor's name to display their details
_date_ - Display all of the councillors by when they started in office either in a date range using , between the dates or a single date
## Committees

_councillor_id_ - Display the committees this councillor is in
## Agendas

_committee_id_ - Display an Agenda by its committee id
## Items

_item_type_id_ - Displays all items with this type id
_agenda_id_ - Displays all items with this agenda id
_councillor_id_ - Displays all items with this councillor id
## Motions

_councillor_id_ - Display all motions by a councillor id
_Item_id_ - Display all motions with this item id
_motion_type_id_ - Display all motions with this motion type id
## Wards

Currently there is no other attributes you can use for this API

SpiritBreaker226 committed:
Add Committee ID Search To API Agenda

SpiritBreaker226 committed:
Move API Paginate From SQL Statement To Render

Move the API paginate from SQL statement because the code was becoming
unmanageable this way it is on one line to do all of the paginate of the
API.

SpiritBreaker226 committed:
Add Item Type ID Search To API Items

SpiritBreaker226 committed:
Change API Item Query To Use Multiple Criteria

SpiritBreaker226 committed:
Move API Order To Render Line In Order Reduce Code

SpiritBreaker226 committed:
Update Item API Query Use Variable Instead Model

SpiritBreaker226 committed:
Change API Motions Query To Use Multiple Criteria

SpiritBreaker226 committed:
Add More Details To Montion API Index

SpiritBreaker226 committed:
Change API Agendas Query To Use Multiple Criteria

SpiritBreaker226 committed:
Change API Agenda To Use @agendas For Searching

Change the API agenda querying to use @agendas for searching in order to
match the reset of the API querying that are using multiple criteria.

SpiritBreaker226 committed:
Fix Missing @motions For Querying API Motions

SpiritBreaker226 committed:
Change API Councillors Query To Multiple Criteria

emilyjfan commented:
Additional queries: 
- Find all the committees for a single councillor
- Find all the councillors that belong to a single committee

cassiascheffer commented:
@SpiritBreaker226 I think what you've done with the API is great! I'm just coming up with ideas here, so let me know if some of these are not possible:

Is it possible to query items by date or date range? e.g. all items between January 1 and January 20, 2014.

Is it possible to query items by words in the title? This would work similar to a search. e.g., All itmes with "subway" in the title.

If we can query items by words in the title, can we also ask for all of the motions on that item?

Since things like ward_id are specific to our database, can we query Councillors by ward_number? So if I ask for the Councillor for Ward 19, I get Mike Layton.
- ward_id and ward number won't necessarily match up.

Can we query motions by Councillor Name rather than Councillor ID? Again, things like councillor_id are specific to our app and the name would be more relevant to other users. 

Similar to the above question, can we query motions by type name, e.g,  "Action", rather than type_id.

SpiritBreaker226 commented:
I'm thinking we need to start documenting how the API as it is getting more complex. I will do the first draft of it, to at the very least write down all of the ways to do searching in the API. Maybe also add this to the ReadMe.md so everyone can see how it works.

@ScottKbka Thank you, I will do my best to do most of was suggested
1. _Is it possible to query items by date or date range? e.g. all items between January 1 and January 20, 2014._
   - Yes, It is possible to do a date range I'm thinking of doing something similar to how orders work with a dash between the dates
2. _Is it possible to query items by words in the title?_
   - Currently, you can do this search using q in the URL for only the item's title.
3. _If we can query items by words in the title, can we also ask for all of the motions on that item?_
   - You can some what do this in motions API side 
4. _Since things like ward_id are specific to our database, can we query Councillors by ward_number?_
   - You can find the councillor by ward number
5. _Can we query motions by Councillor Name rather than Councillor ID?_
   - This is possible and will be done
6. _Can we query motions by type name?_
   - This is possible and will be done

SpiritBreaker226 committed:
Change API Committees Query To Multiple Criteria

SpiritBreaker226 committed:
Add Search By Controller Name In API Controller

SpiritBreaker226 committed:
Add Back Councillor's Tem Items To Test API

SpiritBreaker226 committed:
Add Councillor ID Search for API Item

SpiritBreaker226 committed:
Add Agenda ID Search for API Item

SpiritBreaker226 committed:
Add Date Range To API Councillor For Start Date

SpiritBreaker226 committed:
Add search_by_date_range() To API Controller

Add search_by_date_range() to the API Controller because both the
Councillor model which will use it to search throw start date in office
and Agenda's date in search when agendas are in that range.

Also added in the search date by range from API Controller into API
Councillor

SpiritBreaker226 committed:
Add Search Date Range From API To API Agenda

Add in the Search Date Range method from API Controlloer to API Agenda
in order for Agenda's date to be search in a date range.

SpiritBreaker226 committed:
Remove Councillor's Fake Item Data From Seed

Remove Councillor's fake item data from database seeds because
Councillor's would also be in the Agendas as well as committee items.
